# Laboratorio de Introducción al Procesamiento de Lenguaje Natural 2017

Grupo 20 - Mateo Nogueira - 4.830.584-6

## 1) Importación del corpus
Utilizamos la librería pandas para importar y analizar el corpus de los tweets


In [111]:
import pandas
# Entrenamiento
corpus = pandas.read_csv("corpus_humor_training.csv",encoding='utf-8')
# Validación
corpus2 = pandas.read_csv("corpus_humor_testing.csv",encoding='utf-8')

## 2) Análisis del corpus
Realizamos un breve análisis del contenido del corpus antes del preprocesamiento

In [112]:
print("Entrenamiento \n")

cantTweets = len(corpus.index)

cantVotos1 = corpus['1'].sum()
cantVotos2 = corpus['2'].sum()
cantVotos3 = corpus['3'].sum()
cantVotos4 = corpus['4'].sum()
cantVotos5 = corpus['5'].sum()
cantVotosN = corpus['n'].sum()

print ("El total de tweets es de: " + str(cantTweets)+"\n")
print ("La cantidad de votos de 1 estrella es de: "+str(cantVotos1))
print ("La cantidad de votos de 2 estrella es de: "+str(cantVotos2))
print ("La cantidad de votos de 3 estrella es de: "+str(cantVotos3))
print ("La cantidad de votos de 4 estrella es de: "+str(cantVotos4))
print ("La cantidad de votos de 5 estrella es de: "+str(cantVotos5))
print ("La cantidad de votos negativos es de: "+str(cantVotosN)+"\n")

print ("El total de votos es de: "+str(cantVotos1+cantVotos2+cantVotos3+cantVotos4+cantVotos5+cantVotosN))

# Evaluacion

print("\nEvaluación\n")

cantTweets = len(corpus2.index)

cantVotos1 = corpus2['1'].sum()
cantVotos2 = corpus2['2'].sum()
cantVotos3 = corpus2['3'].sum()
cantVotos4 = corpus2['4'].sum()
cantVotos5 = corpus2['5'].sum()
cantVotosN = corpus2['n'].sum()

print ("El total de tweets es de: " + str(cantTweets)+"\n")
print ("La cantidad de votos de 1 estrella es de: "+str(cantVotos1))
print ("La cantidad de votos de 2 estrella es de: "+str(cantVotos2))
print ("La cantidad de votos de 3 estrella es de: "+str(cantVotos3))
print ("La cantidad de votos de 4 estrella es de: "+str(cantVotos4))
print ("La cantidad de votos de 5 estrella es de: "+str(cantVotos5))
print ("La cantidad de votos negativos es de: "+str(cantVotosN)+"\n")

print ("El total de votos es de: "+str(cantVotos1+cantVotos2+cantVotos3+cantVotos4+cantVotos5+cantVotosN))

Entrenamiento 

El total de tweets es de: 12106

La cantidad de votos de 1 estrella es de: 2960
La cantidad de votos de 2 estrella es de: 2421
La cantidad de votos de 3 estrella es de: 3274
La cantidad de votos de 4 estrella es de: 2541
La cantidad de votos de 5 estrella es de: 1616
La cantidad de votos negativos es de: 14131

El total de votos es de: 26943

Evaluación

El total de tweets es de: 3027

La cantidad de votos de 1 estrella es de: 738
La cantidad de votos de 2 estrella es de: 663
La cantidad de votos de 3 estrella es de: 856
La cantidad de votos de 4 estrella es de: 696
La cantidad de votos de 5 estrella es de: 454
La cantidad de votos negativos es de: 3457

El total de votos es de: 6864


En el conjunto de entrenamiento tenemos 12106 tweets y 26943 votos lo que nos da un promedio de 2,2 votos por tweet mientras que en el de evaluación tenemos 3027 tweets y 3457 votos dando un promedio de 1,1 votos por tweet.

Ambos suman 15133 tweets por lo que se realizó una separacion entre entrenamiento y evaluación de 75%/25%

In [108]:
for text in corpus['text'][140:160]:
    print(text + '\n')

TIpico: Ir por la calle. Saludar a alguien. Que no te devuelva el saludo. Pensar... A ÉSTE YA NO LE SALUDO MÁS.

Decir que es una larga historia para no contarla.

#CancionesParaRumbear Sexy And Know It - LMFAO

—Doctor, doctor, no puedo recordar nada. —¿Desde cuándo tiene este problema? —¿Cuál problema?

La gente lo llama día de San Valentín. Yo lo llamo jueves.

RT @CiseProgresa: El entorno y los consumidores se han transformado completamente y los que estén preparados para adaptarse liderarán el nu…

Por esas abuelas que no aceptan un "no quiero más" como respuesta.

Típico: Ir a la casa de tu amigo y que su madre te tome como ejemplo de buen comportamiento, cuando eres peor que él.

Si te dice fea, es porque eres guapa, nadie le dice fea a una fea.

Mirar el celular cuando no quieres saludar a alguien.

Matar mentalmente a quien no soportas.

Todos tenemos una canción que nos hace llorar, sonreír y recordar. ¿Cuál es la tuya?

—Tienes clase, hijo. —Gracias, mamá, soy un caballero. 

En estos 20 tweets impresos se puede observar como hay distintos tipos de humor. 

Vemos como chistes que utilizan diálogos gracias a los ' - ', algunos tweets que contienen una pregunta y su respuesta, y simples enunciados. En estos 20 no ocurre, pero mirando por arriba todos los tweets detectamos que había tweets informaban de noticias o anunciaban ventas entre otros.

Por ejemplo:

"COMPRA AQUI=&gt;http://t.co/AmFcnsGlM7      

OFERTAS=&gt;http://t.co/9aHOTbR5kP  "

Hay tweets que utilizan otros idiomas para intentar ser humorísticos, por ejemplo el sigiuente tweet contiene letras del alfabeto árabe para decir que los médicos escriben cosas ilegibles. No sabemos que podría suceder al pasarle a un analizador como freeling este tipo de tweets.

"— ¿Entonces qué tengo doctor? — خيف الخاص خي — Gracias!"


Encontramos también un tweet que intenta ser gracioso jugando con la pronunciación. En este caso el clasificador debería ser capaz de detectar que 'digite' puede ser la pronunciación de 'dijiste' para alguna persona que no pronuncie la 's'. Esto claramente supone un problema debido a que sería necesario considerar como se pronuncia el tweet más que solo como esta escrito.

"Un indito en un cajero lee: "Digite la clave"... Y dice: ¡No, yo no li diji la clave a naiden!  #fb"


Finalmente encontramos hashtags y etiquetas de otras cuentas que consideramos que no aportan nada a la clasificación.

Hay que tener en cuenta que hay 12106 tweets y esto fue solamente un breve análisis viendo algunos pocos.

## 3) Preprocesamiento
Como se solicita en la consigna eliminamos los tweets con menos de tres votos y además creamos la columna humoristico con contenido si/no en base a si es humoristico o no. ("Humorístico si la mitad o más de los anotadores lo calificaron con una o más estrellas y no humorístico en caso contrario.")

In [109]:
import numpy as np

# Elimino los tweets con menos de tres votos
corpus = corpus.drop(corpus[corpus['1'] +corpus['2']+corpus['3']+corpus['4']+corpus['5']+corpus['n'] < 3].index)
corpus2 = corpus2.drop(corpus2[corpus2['1'] +corpus2['2']+corpus2['3']+corpus2['4']+corpus2['5']+corpus2['n'] < 3].index)

# Creo la columna humoristico para clasificar a un tweet si cumple la condición o no.
corpus['humoristico'] = np.where(corpus['1']+corpus['2']+corpus['3']+corpus['4']+corpus['5'] >= (corpus['n']+corpus['1']+corpus['2']+corpus['3']+corpus['4']+corpus['5'])/2,'si','no')
corpus2['humoristico'] = np.where(corpus2['1']+corpus2['2']+corpus2['3']+corpus2['4']+corpus2['5'] >= (corpus2['n']+corpus2['1']+corpus2['2']+corpus2['3']+corpus2['4']+corpus2['5'])/2,'si','no')

In [110]:
print("Entrenamiento \n")

cantTweets = len(corpus.index)
print ("El total de tweets es de: " + str(cantTweets)+"\n")

si = 0
no = 0

for index, row in corpus.iterrows():
    if row['humoristico'] == 'si':
        si = si + 1
    else:
        no = no + 1

print ("Cantidad de tweets humoristicos: "+ str(si))
print ("Cantidad de tweets no humoristicos: "+str(no))

print("\nEvaluación\n")

cantTweets = len(corpus2.index)
print ("El total de tweets es de: " + str(cantTweets)+"\n")

si = 0
no = 0

for index, row in corpus2.iterrows():
    if row['humoristico'] == 'si':
        si = si + 1
    else:
        no = no + 1

print ("Cantidad de tweets humoristicos: "+ str(si))
print ("Cantidad de tweets no humoristicos: "+str(no))

Entrenamiento 

El total de tweets es de: 3438

Cantidad de tweets humoristicos: 1835
Cantidad de tweets no humoristicos: 1603

Evaluación

El total de tweets es de: 877

Cantidad de tweets humoristicos: 476
Cantidad de tweets no humoristicos: 401


Podemos ver que la cantidad de tweets bajo bastante, en el caso del conjunto de entrenamiento tenemos más de 3 veces menos tweets y en el de entrenamiento sucede algo similar.

Tenemos para ambos casos mas tweets humorísticos y la cantidad de tweets de evaluación es de aproximadamente (en tamaño) el 25% de la cantidad de entrenamiento.


## 4) Clasificador Parte 1
Para realizar el clasificador utilizaremos la librería scikit-learn. La misma es una librería de aprendizaje automático para python de código abierto que implementa una gran cantidad de algoritmos. En particular implementa los algoritmos de Naïve Bayes y SVM que podrán sernos de utilidad a la hora de clasificar los tweets. 

Para empezar importamos algunas funciones y definimos otras que nos van a ser de utilidad para procesar y evaluar el rendimiento de la clasificación.

In [26]:
from sklearn.model_selection import train_test_split
from sklearn import metrics

from scipy.stats import sem

# Dividimos el corpus en entrenamiento y evaluación. Con una proporción de 80%/20% 
X_train, X_test, y_train, y_test = train_test_split(corpus.text, corpus.humoristico, test_size=0.20, random_state=1)

# La siguiente función se encarga de entrenar un clasificador y 
def train_and_evaluate(clf, X_train, X_test, y_train, y_test):
    clf.fit(X_train, y_train)
    print ("Accuracy en el conjunto de entrenamiento:")
    print (clf.score(X_train, y_train))
    print ("Accuracy en el conjunto de prueba:")
    print (clf.score(X_test, y_test))
    y_pred = clf.predict(X_test)
    print ("Reporte:")
    print (metrics.classification_report(y_test, y_pred))
    print ("Matriz de confusión:")
    print (metrics.confusion_matrix(y_test, y_pred))

Para entrenar un clasificador necesitamos extraer los features primero. scikit-learn provee varias clases para esta tarea. En particular posee un modulo específico para extracción de features en texto. Utilizaremos la clase CountVectorizer. La misma convierte los textos a una matriz númerica.

CountVectorizer funciona siguiendo un estilo bag of words. (preprocesa, tokeniza, cuenta y normaliza). También es posible específicar stop words.

Destacamos que todo esto lo hace automáticamente sin intervención de nosotros. Sin embargo vamos a modificar algunas de estas funciones para tokenizar utilizando Freeling y borrar los hashtags de los tweets.

Vamos a utilizar la clase Pipeline de scikit-learn que nos permite extraer los features y pasarselos a un clasificador en una misma línea, lo cual reduce un poco la cantidad de líneas utilizadas mejorando la legibilidad.

Vamos a probar utilizar un clasificador Naïve Bayes uno Support Vector Machines y uno de entropía máxima.

In [5]:
# Importamos Pipeline
from sklearn.pipeline import Pipeline

# Importamos los clasificadores SVM, NB y entropía máxima
from sklearn.linear_model import SGDClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegressionCV

from sklearn.feature_extraction.text import CountVectorizer

### Tokenización con FreeLing
Lo primero que vamos a realizar es tokenizar nosotros los tweets utilizando FreeLing. Esta es una herramienta para análisis de lenguaje y esta hecha en C++. Trae una API para python.

FreeLing lo bajamos e instalamos desde su página de GitHub
(https://github.com/TALP-UPC/FreeLing/releases)
Para poder ejecutar lo siguiente es necesario tenerlo instalado y configurar las variables FREELINGDIR y DATA

In [6]:
# Inicialización de freeling.

import freeling


FREELINGDIR = "/usr/local";

DATA = FREELINGDIR+"/share/freeling/";
LANG="es";

freeling.util_init_locale("default");

# create language analyzer
la=freeling.lang_ident(DATA+"common/lang_ident/ident.dat");

# create options set for maco analyzer. Default values are Ok, except for data files.
op= freeling.maco_options("es");
op.set_data_files( "", 
                   DATA + "common/punct.dat",
                   DATA + LANG + "/dicc.src",
                   DATA + LANG + "/afixos.dat",
                   "",
                   DATA + LANG + "/locucions.dat", 
                   DATA + LANG + "/np.dat",
                   DATA + LANG + "/quantities.dat",
                   DATA + LANG + "/probabilitats.dat");

# create analyzers
tk=freeling.tokenizer(DATA+LANG+"/tokenizer.dat");
sp=freeling.splitter(DATA+LANG+"/splitter.dat");
sid=sp.open_session();
mf=freeling.maco(op);

# activate mmorpho odules to be used in next call
mf.set_active_options(False, True, True, True,  # select which among created 
                      True, True, False, True,  # submodules are to be used. 
                      True, True, True, True ); # default: all created submodules are used

# create tagger, sense anotator, and parsers
tg=freeling.hmm_tagger(DATA+LANG+"/tagger.dat",True,2);
sen=freeling.senses(DATA+LANG+"/senses.dat");
parser= freeling.chart_parser(DATA+LANG+"/chunker/grammar-chunk.dat");
dep=freeling.dep_txala(DATA+LANG+"/dep_txala/dependences.dat", parser.get_start_symbol());


In [7]:
# La siguiente función se encarga de recibir un string y realizar su procesamiento

def freelTok(lin):

    l = tk.tokenize(lin);
    ls = sp.split(sid,l,True);

    ls = mf.analyze(ls);
    ls = tg.analyze(ls);
    ls = sen.analyze(ls);
    ls = parser.analyze(ls);
    ls = dep.analyze(ls);
    return ls


In [8]:
# Es necesario cerrar la sesión iniciada por freeling
#sp.close_session(sid);

### Creación de vectorizador

Creamos un vectorizador que hereda de CountVectorizer en el que modificamos la función de tokenizar y preprocesamiento, de forma de eliminar los hashtags, simbolos de puntación y considerar stop words.

Vamos a crear nuestra propia función de tokenizar y preprocesar para pasar como parámetro a CountVectorizer.

Primero se van a eliminar hashtags, etiquetas de otras cuentas y símbolos de puntuación. Después se va a tokenizar usando freeling y eliminar stopwords según la lista obtenida de la libería NLTK.

In [10]:
import nltk
# Ejecuto la siguiente línea para bajar el corpus de stopwords de NLTK.
#nltk.download()

from nltk.corpus import stopwords
stopwords1 = stopwords.words('spanish')


In [11]:
import re
import string
from nltk.corpus import stopwords

def AnalizadorTweet(tweet):
    
    # Borro hashtags
    tweet = re.sub(r'\b#.+','',tweet)
    
    # Borro etiquetas
    tweet = re.sub(r'\b@.+','',tweet)
    
    retornar = []
    postFreeling = freelTok(tweet)
    for s in postFreeling:
        ws=s.get_words()
        for w in ws:
            # Elimino stopwords
            if w.get_lemma().lower() not in stopwords1:
                # Elimino símbolos de puntuacion
                if w.get_lemma() not in string.punctuation:
                    retornar.append(w.get_lemma())
                    retornar.append(w.get_tag())
    #print(retornar)
    return retornar


vectorizador = CountVectorizer(analyzer=AnalizadorTweet)

Antes de entrenar vamos a normalizar utilizando TF-IDF (term frequency - inverse document frequency). Este es un peso estadístico para evaluar que tan importante es una palabra para un documento en un corpus.

El valor TF-IDF (importancia) de un token aumenta proporcionalmente a la cantidad de apariciones que un token tiene en un tweet pero se compensa por la frecuencia de aparición en el corpus.

Esto permite tener en cuenta que en general hay palabras más frecuentes que otras.

In [44]:
# Scikit Learn trae una clase para esto
from sklearn.feature_extraction.text import TfidfTransformer
tfidf = TfidfTransformer()

# Creamos los clasificadores:

# Naïve Bayes
nb= Pipeline([('vect', vectorizador),('tfid',tfidf),('clf', MultinomialNB()),])

# SVM
svm= Pipeline([('vect', vectorizador),('tfid',tfidf),('clf', SGDClassifier(max_iter=1000)),])

# Entropía Máxima
maxent= Pipeline([('vect', vectorizador),('tfid',tfidf),('clf', LogisticRegressionCV()),])


Ahora es necesario seleccionar el modelo ajustando los parámetros específicos de cada clasificador. En el teórico vimos dos posibles opciones separar el corpus de entrenamiento en dos corpus (con una proporción 80/20 %) para entrenar sobre el primero, evaluar sobre el segundo y ajustar o utilizar cross-validation donde el corpus se divide en k partes, se entrena sobre k-1, evalua en la restante, se repite para cada parte y se toma la media.

ScikitLearn posee la clase GridSearchCV que nos permite indicar valores para ciertos parámetros y luego se entrena como un clasificador normal, durante el entrenamiento utiliza cross validation para seleccionar de los parámetros dados los mejores.

En general, para el vectorizador y el tfidf vamos a probar con los siguientes parámetros:

'binary' del vectorizador en True o en False, este parámetro según la documentación de scikitlearn da mejores resultados en True para eventos binarios como el que nos enfrentamos (si / no)

'ngram_range' del vectorizador: Es una tupla (min_n, max_n) que indica el mínimo n y el máximo n de los n-gramas a utilizar. Vamos a probar con unigramas, bigramas, unigramas + bigramas, unigramas + bigramas + trigramas, bigramas + trigramas.

'use_idf' del tfidf: Booleano que permite la reasignación de pesos idf.

'sublinear_tf' del tfidf: Aplica escalamiento sublineal (reemplazar la frecuencia de un término por 1+log(frecuencia))

In [39]:
from sklearn.model_selection import GridSearchCV

En el caso de naïve bayes vamos a probar solamente ajustar los parámetros 'fit_prior' y 'alpha'.


In [48]:
parametros_nb = {'vect__binary': (True, False), 
                 'vect__ngram_range': [(1, 1), (2, 2), (1, 2), (1, 3), (2, 3)],
                 'tfid__use_idf': (True, False),
                 'tfid__sublinear_tf': (True, False),
                 'clf__alpha': (1e-2, 1e-3, 1, 1e-1),
                 'clf__fit_prior': (True, False) }

# n_jobs indica que utilice todos los cores del CPU
grid_nb = GridSearchCV(nb, parametros_nb, n_jobs=-1)

In [43]:
print('Naïve Bayes \n')
# Entreno y evalúo performance
train_and_evaluate(grid_nb,corpus['text'],corpus2['text'],corpus['humoristico'],corpus2['humoristico'])

Naïve Bayes 

Accuracy en el conjunto de entrenamiento:
0.885689354276
Accuracy en el conjunto de prueba:
0.743443557583
Reporte:
             precision    recall  f1-score   support

         no       0.71      0.74      0.73       401
         si       0.77      0.75      0.76       476

avg / total       0.74      0.74      0.74       877

Matriz de confusión:
[[297 104]
 [121 355]]


En el caso de SVM vamos a probar ajustar 'alpha', parámetro que multiplica el término de normalización.

In [47]:
parametros_svm = {'vect__binary': (True, False), 
                 'vect__ngram_range': [(1, 1), (2, 2), (1, 2), (1, 3), (2, 3)],
                 'tfid__use_idf': (True, False),
                 'tfid__sublinear_tf': (True, False),
                 'clf__alpha': (1e-2, 1e-4, 1e-1) }

# n_jobs indica que utilice todos los cores del CPU

grid_svm = GridSearchCV(svm, parametros_svm, n_jobs=-1)

In [41]:
print('SVM \n')
# Entreno y evalúo performance
train_and_evaluate(grid_svm,corpus['text'],corpus2['text'],corpus['humoristico'],corpus2['humoristico'])

SVM 

Accuracy en el conjunto de entrenamiento:
0.88801628854
Accuracy en el conjunto de prueba:
0.745724059293
Reporte:
             precision    recall  f1-score   support

         no       0.73      0.71      0.72       401
         si       0.76      0.78      0.77       476

avg / total       0.75      0.75      0.75       877

Matriz de confusión:
[[283 118]
 [105 371]]


In [50]:
parametros_maxent = {'vect__binary': (True, False), 
                     'vect__ngram_range': [(1, 1), (2, 2), (1, 2), (1, 3), (2, 3)],
                     'tfid__use_idf': (True, False),
                     'tfid__sublinear_tf': (True, False),
                     'clf__refit': (True, False),
                     'clf__fit_intercept': (True, False) }

# n_jobs indica que utilice todos los cores del CPU
grid_maxent = GridSearchCV(maxent, parametros_maxent, n_jobs=-1)

In [45]:
print('Máxima Entropía \n')
# Entreno y evalúo performance
train_and_evaluate(grid_maxent,corpus['text'],corpus2['text'],corpus['humoristico'],corpus2['humoristico'])

Máxima Entropía 

Accuracy en el conjunto de entrenamiento:
0.837696335079
Accuracy en el conjunto de prueba:
0.762827822121
Reporte:
             precision    recall  f1-score   support

         no       0.74      0.75      0.74       401
         si       0.79      0.77      0.78       476

avg / total       0.76      0.76      0.76       877

Matriz de confusión:
[[301 100]
 [108 368]]


Se puede observar que naive bayes y SVM tuvieron, en promedio, un accuracy similar tanto en el conjunto de prueba como en el de entrenamiento de aproximadamente 74%. Ambos alcanzan casi un 88% en el conjunto de entrenamiento lo que muestra que los clasificadores estan teniendo un sobreajuste sobre en el conjunto de entrenamiento debido a la diferencia con el desempeño en el de prueba.

El clasificador de entropía máxima no solo obtuvo un desempeño en el conjunto de prueba si no que además la diferencia con el conjunto de entrenamiento es mucho menor que en los anteriores, lo que indica menos sobreajuste. En este caso la matriz de confusión nos muestra que se equivoco la misma cantidad de tweets.

Los tres tuvieron una mejor precisión cuando se trato de tweets humorísticos, por lo que estan un poco desvíados hacía el 'si'. En el caso de Naïve Bayes, la diferencia de precisión entre 'si' y 'no' alcanzo un 6%, en SVM un 3% y en entropía máxima un 5%. En cuanto a recall se observan porcentajes similares a los de la precisión por lo que la medida f1, siendo una medida armónica entre precisión y recall también es similar a la precisión y el recall. 


En base a lo anterior consideramos que el mejor clasificador fue el de entropía máxima. 

## 4) Clasificador Parte 2
Para realizar este clasificador necesitamos ademas estimar la mediana de los tweets humorísticos. Para esto creamos una nueva columna 'mediana' de la misma manera que antes solo que ahora en ves de 'si' va a ir la mediana de los votos y 0 va a significar no humorístico.

In [51]:
from statistics import median
from math import ceil

mediana = []

for index, row in corpus.iterrows():
    Calcmediana = []
    
    for iter in range (0, int(row['1'])):
        Calcmediana.append(1)
    for iter in range (0, int(row['2'])):
        Calcmediana.append(2)
    for iter in range (0, int(row['3'])):
        Calcmediana.append(3)
    for iter in range (0, int(row['4'])):
        Calcmediana.append(4)
    for iter in range (0, int(row['5'])):
        Calcmediana.append(5)
    for iter in range (0, int(row['n'])):
        Calcmediana.append(0)
        
    mediana.append(ceil(median(Calcmediana)))
    
mediana2 = []

for index, row in corpus2.iterrows():
    Calcmediana2 = []
    
    for iter in range (0, int(row['1'])):
        Calcmediana2.append(1)
    for iter in range (0, int(row['2'])):
        Calcmediana2.append(2)
    for iter in range (0, int(row['3'])):
        Calcmediana2.append(3)
    for iter in range (0, int(row['4'])):
        Calcmediana2.append(4)
    for iter in range (0, int(row['5'])):
        Calcmediana2.append(5)
    for iter in range (0, int(row['n'])):
        Calcmediana2.append(0)
        
    mediana2.append(ceil(median(Calcmediana2)))


corpus2['mediana'] = mediana2    
corpus['mediana'] = mediana

Vuelvo a entrenar los clasificadores

In [52]:
print('Naïve Bayes \n')
# Entreno y evalúo performance
train_and_evaluate(grid_nb,corpus['text'],corpus2['text'],corpus['mediana'],corpus2['mediana'])

Naïve Bayes 

Accuracy en el conjunto de entrenamiento:
0.749854566608
Accuracy en el conjunto de prueba:
0.476624857469
Reporte:
             precision    recall  f1-score   support

          0       0.55      0.94      0.69       401
          1       0.27      0.10      0.14       125
          2       0.23      0.13      0.17       122
          3       0.18      0.11      0.14       139
          4       0.00      0.00      0.00        77
          5       0.00      0.00      0.00        13

avg / total       0.35      0.48      0.38       877

Matriz de confusión:
[[375   6  10   9   1   0]
 [ 87  12   9  17   0   0]
 [ 78   8  16  20   0   0]
 [ 95  11  18  15   0   0]
 [ 39   6  16  16   0   0]
 [  5   1   2   5   0   0]]


/home/mateo/.local/lib/python3.5/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [53]:
print('SVM \n')
# Entreno y evalúo performance
train_and_evaluate(grid_svm,corpus['text'],corpus2['text'],corpus['mediana'],corpus2['mediana'])

SVM 

Accuracy en el conjunto de entrenamiento:
0.524432809773
Accuracy en el conjunto de prueba:
0.45724059293
Reporte:
             precision    recall  f1-score   support

          0       0.50      0.93      0.65       401
          1       0.28      0.07      0.11       125
          2       0.19      0.04      0.07       122
          3       0.22      0.09      0.13       139
          4       0.06      0.01      0.02        77
          5       0.00      0.00      0.00        13

avg / total       0.34      0.46      0.35       877

Matriz de confusión:
[[373   6   8  10   3   1]
 [ 96   9   6  12   2   0]
 [ 91   5   5  15   4   2]
 [109   6   7  13   4   0]
 [ 62   5   1   8   1   0]
 [  8   1   0   2   2   0]]


In [54]:
print('Entropía Máxima \n')
# Entreno y evalúo performance
train_and_evaluate(grid_maxent,corpus['text'],corpus2['text'],corpus['mediana'],corpus2['mediana'])

Entropía Máxima 

Accuracy en el conjunto de entrenamiento:
0.4845840605
Accuracy en el conjunto de prueba:
0.467502850627
Reporte:
             precision    recall  f1-score   support

          0       0.50      0.97      0.66       401
          1       0.21      0.18      0.19       125
          2       0.00      0.00      0.00       122
          3       0.00      0.00      0.00       139
          4       0.00      0.00      0.00        77
          5       0.00      0.00      0.00        13

avg / total       0.26      0.47      0.33       877

Matriz de confusión:
[[388  13   0   0   0   0]
 [103  22   0   0   0   0]
 [ 96  26   0   0   0   0]
 [115  24   0   0   0   0]
 [ 62  15   0   0   0   0]
 [  7   6   0   0   0   0]]


/home/mateo/.local/lib/python3.5/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


Obtuvimos resultados similares en el conjunto de evaluación para los tres clasificadores. No sucede lo mismo en el conjunto de entrenamiento donde en Naïve Bayes se observa un sobreajuste al corpus de entrenamiento debido a que hay una diferencia de casi 30% entre el accuracy en evaluación y entrenamiento.

Mirando la precisión, vemos que el clasificador de entropía máxima fue el peor de todos, dado que no clasifico ningún tweet como de dos o más estrellas ni correctamente ni incorrectamente. La mejor precisión la tuvo el clasificador de SVM, que fue capaz de clasificar bien tweets con 4 estrellas, algo que los otros dos no pudieron.

Hay un recall muy alto en 0 estrellas de 93%, 94% y 97% y considerando su precisión de 50%, 50% y 55% nos muestra que hay una tendencia a marcar un tweet como 0 estrellas.


Decidimos por el motivo de poder detectar tweets de hasta 4 estrellas que el mejor clasificador en esta etapa es del SVM.

Tenemos que observar también que se observan warnings de que hay etiquetas sin muestras predecidas, por lo que consideramos que esta clasificación no es muy fiable y deberia repetirse la prueba con un corpus más grande, con más ejemplos para todas las etiquetas.



## 5) Clasificador Parte 3
Para este clasificador vamos a utilizar la columna del corpus 'mediana' calculada en la parte anterior. En base a esta columna creamos una nueva 'humoristicoP3' que va a valer 'si' sí para esa fila la columna 'mediana' valia 1 o más y 'no' en caso de que 'mediana' tuviera valor 0.

In [55]:
humoristicoP3 = []
humoristicoP3_2 = []

for index, row in corpus.iterrows():
    if row['mediana'] > 0:
        humoristicoP3.append('si')
    else:
        humoristicoP3.append('no')
        
for index, row in corpus2.iterrows():
    if row['mediana'] > 0:
        humoristicoP3_2.append('si')
    else:
        humoristicoP3_2.append('no')
        

corpus['humoristicoP3'] = humoristicoP3
corpus2['humoristicoP3'] = humoristicoP3_2

In [56]:
print('Naïve Bayes \n')
# Entreno y evalúo performance
train_and_evaluate(grid_nb,corpus['text'],corpus2['text'],corpus['humoristicoP3'],corpus2['humoristicoP3'])

Naïve Bayes 

Accuracy en el conjunto de entrenamiento:
0.885689354276
Accuracy en el conjunto de prueba:
0.743443557583
Reporte:
             precision    recall  f1-score   support

         no       0.71      0.74      0.73       401
         si       0.77      0.75      0.76       476

avg / total       0.74      0.74      0.74       877

Matriz de confusión:
[[297 104]
 [121 355]]


In [57]:
print('SVM \n')
# Entreno y evalúo performance
train_and_evaluate(grid_svm,corpus['text'],corpus2['text'],corpus['humoristicoP3'],corpus2['humoristicoP3'])

SVM 

Accuracy en el conjunto de entrenamiento:
0.888307155323
Accuracy en el conjunto de prueba:
0.745724059293
Reporte:
             precision    recall  f1-score   support

         no       0.73      0.71      0.72       401
         si       0.76      0.78      0.77       476

avg / total       0.75      0.75      0.75       877

Matriz de confusión:
[[283 118]
 [105 371]]


In [58]:
print('Entropía Máxima \n')
# Entreno y evalúo performance
train_and_evaluate(grid_maxent,corpus['text'],corpus2['text'],corpus['humoristicoP3'],corpus2['humoristicoP3'])

Entropía Máxima 

Accuracy en el conjunto de entrenamiento:
0.837696335079
Accuracy en el conjunto de prueba:
0.762827822121
Reporte:
             precision    recall  f1-score   support

         no       0.74      0.75      0.74       401
         si       0.79      0.77      0.78       476

avg / total       0.76      0.76      0.76       877

Matriz de confusión:
[[301 100]
 [108 368]]


Obtuvimos resultados exactamente iguales a los del clasificador de la parte 1. Vamos a ver si hubo algún tweet que cambio de humorístico a no humorístico comparando ambos criterios.

In [106]:
for index, row in corpus.iterrows():
    if row['humoristico'] != row['humoristicoP3']:
        print(row['text']+"\n")

La salida en blanco muestra que en este corpus ambos criterios son equivalentes.

## 6) Conclusiones

De los resultados anteriores podemos concluir que el criterio utilizado para la parte 1 y  la parte 3 nos dan resultados equivalentes. Esto es, considerar un tweet como humorístico si la mitad o más de los anotadores lo calificaron con una o más estrellas, o, si la mediana de la cantidad de estrellas es mayor que 0 son equivalentes. Para un trabajo futuro se podría intentar encontrar un contraejemplo para mostrar si esto ocurre siempre o fue casualidad o en caso de no encontrarlo intentar demostrar matemáticamente que son equivalentes.

Para la parte 2 obtuvimos resultados muy malos en comparación con las otras partes pero consideramos que es debido a que tenemos muchas etiquetas y pocas muestras para cada una por lo que es necesario obtener un corpus más grande con la suficiente cantidad de muestras clasificadas y repetir el experimento en un futuro.

También es necesario tener en cuenta que el humor es subjetivo y lo que para una persona es humorístico para otra podría no serlo. Se podría ser un poco mas específico en este tema si se distinguiera por tipo de chiste que contiene el tweet o quienes son los anotadores (un ejemplo para esto sería distingiuir por franjas de edades)[1].

[1] https://www.fing.edu.uy/inco/grupos/pln/prygrado/Informepghumor.pdf

[2] http://scikit-learn.org/stable/documentation.html

[3] http://nlp.lsi.upc.edu/freeling/

[4] http://www.nltk.org/
